// @ng/common/dist/es2015/src/interceptors/progressInterceptor/progressInterceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { HttpEventType, HTTP_INTERCEPTORS } from '@angular/common/http';
import { tap } from 'rxjs/operators';
import { ProgressIndicatorService } from '../../components/progressIndicator/progressIndicator.service';
import { IgnoredInterceptorsService } from '../../services/ignoredInterceptors/ignoredInterceptors.service';
/**
 * ProgressInterceptor used for intercepting http requests and displaying progress indicatior
 */
export class ProgressInterceptor {
    //######################### constructors #########################
    /**
     * @param {?} _indicatorSvc
     * @param {?=} _ignoredInterceptorsService
     */
    constructor(_indicatorSvc, _ignoredInterceptorsService) {
        this._indicatorSvc = _indicatorSvc;
        this._ignoredInterceptorsService = _ignoredInterceptorsService;
    }
    //######################### public methods - implementation of HttpInterceptor #########################
    /**
     * Intercepts http request
     * @param {?} req Request to be intercepted
     * @param {?} next Next middleware that can be called for next processing
     * @return {?}
     */
    intercept(req, next) {
        return next.handle(req)
            .pipe(tap((/**
         * @param {?} event
         * @return {?}
         */
        event => {
            if (this._ignoredInterceptorsService && this._ignoredInterceptorsService.isIgnored(ProgressInterceptor, req)) {
                return;
            }
            //request started
            if (event.type == HttpEventType.Sent) {
                this._indicatorSvc.showProgress();
            }
            //response received
            else if (event.type == HttpEventType.Response) {
                this._indicatorSvc.hideProgress();
            }
        }), (/**
         * @return {?}
         */
        () => this._indicatorSvc.hideProgress())));
    }
}
ProgressInterceptor.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ProgressInterceptor.ctorParameters = () => [
    { type: ProgressIndicatorService },
    { type: IgnoredInterceptorsService, decorators: [{ type: Optional }] }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ProgressInterceptor.prototype._indicatorSvc;
    /**
     * @type {?}
     * @private
     */
    ProgressInterceptor.prototype._ignoredInterceptorsService;
}
/**
 * Progress interceptor provider
 * @type {?}
 */
export const PROGRESS_INTERCEPTOR_PROVIDER = {
    provide: HTTP_INTERCEPTORS,
    useClass: ProgressInterceptor,
    multi: true
};
