// @ng/common/dist/es2015/src/components/progressIndicator/progressIndicator.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ProgressIndicatorService } from './progressIndicator.service';
/**
 * Component that is used for displaying progress indicator
 */
export class ProgressIndicatorComponent {
    //######################### constructor #########################
    /**
     * @param {?} _service
     */
    constructor(_service) {
        this._service = _service;
        //######################### public properties - inputs #########################
        /**
         * Applied css classes
         */
        this.appliedClass = { "progress-indicator": true };
        this.running = this._service.running;
        this._subscription = this._service.runningChanged.subscribe((/**
         * @param {?} running
         * @return {?}
         */
        running => this.running = running));
    }
    /**
     * Sets css classes that will be applied to indicator
     * @param {?} cssClass
     * @return {?}
     */
    set cssClass(cssClass) {
        this.appliedClass = {};
        this.appliedClass[cssClass] = true;
    }
    //######################### private properties #########################
    /**
     * Sets indication that progress indicator is running
     * @private
     * @param {?} running
     * @return {?}
     */
    set running(running) {
        this.appliedClass["running"] = running;
    }
    ;
    //######################### public methods - implementation of OnDestroy #########################
    /**
     * Called when component is destroyed
     * @return {?}
     */
    ngOnDestroy() {
        if (this._subscription) {
            this._subscription.unsubscribe();
            this._subscription = null;
        }
    }
}
ProgressIndicatorComponent.decorators = [
    { type: Component, args: [{
                selector: "progress-indicator",
                template: "<div [ngClass]=\"appliedClass\">\r\n    <div></div>\r\n</div>",
                styles: ["\r\n.progress-indicator\r\n{\r\n    -webkit-transition: all 250ms linear;\r\n    -moz-transition: all 250ms linear;\r\n    transition: all 250ms linear;\r\n    background-color: rgba(0, 0, 0, 0.3);\r\n    height: 100%;\r\n    width: 100%;\r\n    position: fixed;\r\n    top: 0;\r\n    left: 0;\r\n    z-index: 23456;\r\n    opacity: 0;\r\n    visibility: hidden;\r\n}\r\n.progress-indicator.running\r\n{\r\n    visibility: visible;\r\n    opacity: 1;\r\n}\r\n.progress-indicator.running > div\r\n{\r\n    border-left: 4px solid #eee;\r\n    border-radius: 25px;\r\n    border-top: 4px solid #eee;\r\n    height: 40px;\r\n    margin-left: auto;\r\n    margin-right: auto;\r\n    position: relative;\r\n    top: 50%;\r\n    width: 40px;\r\n    animation-name: rotate-progress;\r\n    animation-duration: 350ms;\r\n    animation-iteration-count: infinite;\r\n    animation-timing-function: linear;\r\n}\r\n\r\n@keyframes rotate-progress\r\n{\r\n    from\r\n    {\r\n        transform: rotate(0deg);\r\n    }\r\n    to\r\n    {\r\n        transform: rotate(360deg);\r\n    }\r\n}"]
            }] }
];
/** @nocollapse */
ProgressIndicatorComponent.ctorParameters = () => [
    { type: ProgressIndicatorService }
];
ProgressIndicatorComponent.propDecorators = {
    cssClass: [{ type: Input }]
};
if (false) {
    /**
     * Subscription for changes in ProgressIndicatorService
     * @type {?}
     * @private
     */
    ProgressIndicatorComponent.prototype._subscription;
    /**
     * Applied css classes
     * @type {?}
     */
    ProgressIndicatorComponent.prototype.appliedClass;
    /**
     * @type {?}
     * @private
     */
    ProgressIndicatorComponent.prototype._service;
    /* Skipping unhandled member: ;*/
}
