// @ng/authentication/dist/es5/src/suppressAuthInterceptor/suppressAuthInterceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { IgnoredInterceptorsService } from '@ng/common';
import { Observable } from 'rxjs';
import { catchError } from 'rxjs/operators';
/**
 * SuppressAuthInterceptor used for intercepting http responses and suppressing 401, 403 statuses
 */
var SuppressAuthInterceptor = /** @class */ (function () {
    //######################### constructors #########################
    function SuppressAuthInterceptor(_ignoredInterceptorsService) {
        this._ignoredInterceptorsService = _ignoredInterceptorsService;
    }
    //######################### public methods - implementation of HttpInterceptor #########################
    /**
     * Intercepts http request
     * @param req Request to be intercepted
     * @param next Next middleware that can be called for next processing
     */
    //######################### public methods - implementation of HttpInterceptor #########################
    /**
     * Intercepts http request
     * @param {?} req Request to be intercepted
     * @param {?} next Next middleware that can be called for next processing
     * @return {?}
     */
    SuppressAuthInterceptor.prototype.intercept = 
    //######################### public methods - implementation of HttpInterceptor #########################
    /**
     * Intercepts http request
     * @param {?} req Request to be intercepted
     * @param {?} next Next middleware that can be called for next processing
     * @return {?}
     */
    function (req, next) {
        var _this = this;
        return next.handle(req).pipe(catchError((/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            return (/** @type {?} */ (Observable.create((/**
             * @param {?} observer
             * @return {?}
             */
            function (observer) {
                //client error, not response from server, or is ignored
                if (err.error instanceof Error ||
                    (_this._ignoredInterceptorsService && _this._ignoredInterceptorsService.isIgnored(SuppressAuthInterceptor, req))) {
                    observer.error(err);
                    observer.complete();
                    return;
                }
                //if auth error
                if (err.status == 403 || err.status == 401) {
                    observer.complete();
                    return;
                }
                //other errors
                observer.error(err);
                observer.complete();
            }))));
        })));
    };
    SuppressAuthInterceptor.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SuppressAuthInterceptor.ctorParameters = function () { return [
        { type: IgnoredInterceptorsService }
    ]; };
    return SuppressAuthInterceptor;
}());
export { SuppressAuthInterceptor };
if (false) {
    /**
     * @type {?}
     * @private
     */
    SuppressAuthInterceptor.prototype._ignoredInterceptorsService;
}
/**
 * Provider for proper use of SuppressAuthInterceptor, use this provider to inject this interceptor
 * @type {?}
 */
export var SUPPRESS_AUTH_INTERCEPTOR_PROVIDER = {
    provide: HTTP_INTERCEPTORS,
    multi: true,
    useClass: SuppressAuthInterceptor
};
