// @ng/select/dist/es2015/src/plugins/valueHandler/basic/basicValueHandler.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Inject, Optional, ElementRef } from '@angular/core';
import { extend, isBlank } from '@asseco/common';
import { NG_SELECT_PLUGIN_INSTANCES } from '../../../components/select/types';
import { VALUE_HANDLER_OPTIONS } from '../types';
import { ValueHandlerBase } from '../valueHandlerBase';
/**
 * Default options for value handler
 * \@internal
 * @type {?}
 */
const defaultOptions = {};
/**
 * Component used for handling current value of NgSelect
 * @template TValue
 */
export class BasicValueHandlerComponent extends ValueHandlerBase {
    //######################### constructor #########################
    /**
     * @param {?} ngSelectPlugins
     * @param {?} pluginElement
     * @param {?=} options
     */
    constructor(ngSelectPlugins, pluginElement, options) {
        super(ngSelectPlugins, pluginElement);
        //######################### protected methods #########################
        /**
         * Sets value
         */
        this._setValue = (/**
         * @param {?} option
         * @return {?}
         */
        (option) => {
            //multiple values are allowed
            if (this.options.multiple) {
                if (!Array.isArray(this.selectedOptions)) {
                    this.selectedOptions = [];
                }
                if (Array.isArray(this.selectedOptions)) {
                    /** @type {?} */
                    let index;
                    //value exists, removing from list
                    if ((index = this.selectedOptions.indexOf(option)) >= 0) {
                        this.selectedOptions.splice(index, 1);
                    }
                    //adding value
                    else {
                        this.selectedOptions.push(option);
                    }
                    this.selectedOptions = [...this.selectedOptions];
                }
            }
            else 
            //only signle value allowed
            {
                this.selectedOptions = option;
            }
            this._clearSelected();
            this._markValueAsSelected();
            this._normalState.invalidateVisuals();
            this.valueChange.emit();
            //close popup if not multiple
            if (!this.options.multiple) {
                this.popupVisibilityRequest.emit(false);
            }
            else {
                this._popup.invalidateVisuals();
            }
        });
        this._options = extend(true, {}, defaultOptions, options);
    }
    //######################### public properties - implementation of BasicValueHandler #########################
    /**
     * Options for NgSelect plugin
     * @return {?}
     */
    get options() {
        return this._options;
    }
    /**
     * @param {?} options
     * @return {?}
     */
    set options(options) {
        this._options = extend(true, this._options, options);
    }
    //######################### public methods - implementation of BasicValueHandler #########################
    /**
     * Sets value for NgSelect
     * @param {?} value Value to be set
     * @return {?}
     */
    setValue(value) {
        this._useOptionsAsValue(value);
    }
    /**
     * Initialize plugin options, all operations required to be done with plugin options are handled here
     * @return {?}
     */
    initOptions() {
    }
    /**
     * Explicitly runs invalidation of content (change detection)
     * @return {?}
     */
    invalidateVisuals() {
    }
    /**
     * Loads options
     * @protected
     * @return {?}
     */
    _loadOptions() {
        this._useOptionsAsValue(isBlank(this._unmappedValue) ? this.value : this._unmappedValue);
    }
    /**
     * Converts value to options
     * @protected
     * @param {?} value Value to be changed to options
     * @return {?}
     */
    _useOptionsAsValue(value) {
        //set empty value
        if (isBlank(value) || (Array.isArray(value) && !value.length)) {
            this.selectedOptions = value;
            this._clearSelected();
            this._unmappedValue = null;
            this._normalState.invalidateVisuals();
            this.valueChange.emit();
            return;
        }
        //no options available yet
        if (!this.optionsGatherer.options || !this.optionsGatherer.options.length) {
            this._unmappedValue = value;
            return;
        }
        if (this.options.multiple) {
            if (Array.isArray(value)) {
                /** @type {?} */
                let items = value;
                this.selectedOptions = this.optionsGatherer.options.filter((/**
                 * @param {?} itm
                 * @return {?}
                 */
                itm => !!items.find((/**
                 * @param {?} it
                 * @return {?}
                 */
                it => this.valueComparer(it, itm.value)))));
            }
            else {
                throw new Error('Don`t you have redundant "multiple"?');
            }
        }
        else {
            if (Array.isArray(value)) {
                throw new Error('Are you missing attribute "multiple"?');
            }
            else {
                /** @type {?} */
                let item = value;
                this.selectedOptions = this.optionsGatherer.options.find((/**
                 * @param {?} itm
                 * @return {?}
                 */
                itm => this.valueComparer(itm.value, item)));
            }
        }
        this._clearSelected();
        this._markValueAsSelected();
        this._unmappedValue = null;
        this._normalState.invalidateVisuals();
        this.valueChange.emit();
    }
}
BasicValueHandlerComponent.decorators = [
    { type: Component, args: [{
                selector: "ng-basic-value-handler",
                template: '',
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
BasicValueHandlerComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [NG_SELECT_PLUGIN_INSTANCES,] }, { type: Optional }] },
    { type: ElementRef },
    { type: undefined, decorators: [{ type: Inject, args: [VALUE_HANDLER_OPTIONS,] }, { type: Optional }] }
];
if (false) {
    /**
     * Backed up unmapped value that was set before options were obtained
     * @type {?}
     * @protected
     */
    BasicValueHandlerComponent.prototype._unmappedValue;
    /**
     * Sets value
     * @type {?}
     * @protected
     */
    BasicValueHandlerComponent.prototype._setValue;
}
