// @ng/select/dist/es5/src/plugins/textsLocator/no/noTextsLocator.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, ElementRef, EventEmitter, Inject, Optional } from "@angular/core";
import { NG_SELECT_PLUGIN_INSTANCES } from '../../../components/select/types';
/**
 * Component used for rendering no texts locator
 */
var NoTextsLocatorComponent = /** @class */ (function () {
    //######################### constructor #########################
    function NoTextsLocatorComponent(ngSelectPlugins, pluginElement) {
        this.ngSelectPlugins = ngSelectPlugins;
        this.pluginElement = pluginElement;
        /**
         * Indication that texts should be obtained again, because they have changed
         */
        this.textsChange = new EventEmitter();
    }
    //######################### public methods - implementation of TextsLocator #########################
    /**
     * Gets text for specified key
     */
    //######################### public methods - implementation of TextsLocator #########################
    /**
     * Gets text for specified key
     * @param {?} key
     * @return {?}
     */
    NoTextsLocatorComponent.prototype.getText = 
    //######################### public methods - implementation of TextsLocator #########################
    /**
     * Gets text for specified key
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return key;
    };
    /**
     * Initialize plugin, to be ready to use, initialize communication with other plugins
     */
    /**
     * Initialize plugin, to be ready to use, initialize communication with other plugins
     * @return {?}
     */
    NoTextsLocatorComponent.prototype.initialize = /**
     * Initialize plugin, to be ready to use, initialize communication with other plugins
     * @return {?}
     */
    function () {
    };
    /**
     * Initialize plugin options, all operations required to be done with plugin options are handled here
     */
    /**
     * Initialize plugin options, all operations required to be done with plugin options are handled here
     * @return {?}
     */
    NoTextsLocatorComponent.prototype.initOptions = /**
     * Initialize plugin options, all operations required to be done with plugin options are handled here
     * @return {?}
     */
    function () {
    };
    /**
     * Explicitly runs invalidation of content (change detection)
     */
    /**
     * Explicitly runs invalidation of content (change detection)
     * @return {?}
     */
    NoTextsLocatorComponent.prototype.invalidateVisuals = /**
     * Explicitly runs invalidation of content (change detection)
     * @return {?}
     */
    function () {
    };
    NoTextsLocatorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-no-texts-locator',
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    NoTextsLocatorComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [NG_SELECT_PLUGIN_INSTANCES,] }, { type: Optional }] },
        { type: ElementRef }
    ]; };
    return NoTextsLocatorComponent;
}());
export { NoTextsLocatorComponent };
if (false) {
    /**
     * Options for NgSelect plugin
     * @type {?}
     */
    NoTextsLocatorComponent.prototype.options;
    /**
     * Indication that texts should be obtained again, because they have changed
     * @type {?}
     */
    NoTextsLocatorComponent.prototype.textsChange;
    /** @type {?} */
    NoTextsLocatorComponent.prototype.ngSelectPlugins;
    /** @type {?} */
    NoTextsLocatorComponent.prototype.pluginElement;
}
