// @ng/common/dist/es2015/src/services/ignoredInterceptors/ignoredInterceptors.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from "@angular/core";
import * as i0 from "@angular/core";
/**
 * Contains id of request, used for IgnoredInterceptorsService
 * @record
 */
export function IgnoredInterceptorId() { }
if (false) {
    /**
     * Identification of request
     * @type {?|undefined}
     */
    IgnoredInterceptorId.prototype.requestId;
}
/**
 * Http request with request id for IgnoredInterceptorsService
 * @record
 * @template TBody
 */
export function HttpRequestIgnoredInterceptorId() { }
/**
 * Service used for disabling specific http client interceptor for one call, if you add interceptor you want to disable before call it will be disabled
 */
export class IgnoredInterceptorsService {
    constructor() {
        //######################### private fields #########################
        /**
         * Array of interceptors that will be ignored
         */
        this._ignoredInterceptors = [];
    }
    //######################### public methods #########################
    /**
     * Clears all ignored interceptors from service
     * @return {?}
     */
    clear() {
        this._ignoredInterceptors = [];
    }
    /**
     * Adds interceptor type that should be ignored for specified url
     * @template TType
     * @param {?} interceptorType Type of interceptor should be ignored
     * @param {?} requestId Object containing request id
     * @return {?}
     */
    addInterceptor(interceptorType, requestId) {
        if (!requestId.requestId) {
            return;
        }
        if (!this._ignoredInterceptors.find((/**
         * @param {?} itm
         * @return {?}
         */
        itm => itm.type == interceptorType && itm.requestId == requestId.requestId))) {
            this._ignoredInterceptors.push({
                type: interceptorType,
                requestId: requestId.requestId
            });
        }
    }
    /**
     * Checks specified interceptor whether is ingored
     * @template TType
     * @param {?} interceptorType Type of interceptor that is checked whether is ignored
     * @param {?} requestId
     * @return {?}
     */
    isIgnored(interceptorType, requestId) {
        if (!requestId.requestId) {
            return false;
        }
        /** @type {?} */
        let item = this._ignoredInterceptors.find((/**
         * @param {?} itm
         * @return {?}
         */
        itm => itm.requestId == requestId.requestId && itm.type == interceptorType));
        if (item) {
            this._ignoredInterceptors.splice(this._ignoredInterceptors.indexOf(item), 1);
            return true;
        }
        return false;
    }
}
IgnoredInterceptorsService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */ IgnoredInterceptorsService.ngInjectableDef = i0.defineInjectable({ factory: function IgnoredInterceptorsService_Factory() { return new IgnoredInterceptorsService(); }, token: IgnoredInterceptorsService, providedIn: "root" });
if (false) {
    /**
     * Array of interceptors that will be ignored
     * @type {?}
     * @private
     */
    IgnoredInterceptorsService.prototype._ignoredInterceptors;
}
