// @ng/select/dist/es5/src/elements/components/ngSelectElements.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Public API for NgSelect WebComponent
 * @record
 * @template TValue
 */
export function NgSelectWebComponent() { }
if (false) {
    /**
     * Occurs every time when NgSelect is initialized or reinitialized, if value is false NgSelect was not initialized yet
     * @type {?}
     */
    NgSelectWebComponent.prototype.initializedWC;
    /**
     * Gets current state of initialization
     * @type {?}
     */
    NgSelectWebComponent.prototype.isInitializedWC;
    /**
     * Gets or sets NgSelect options
     * @type {?}
     */
    NgSelectWebComponent.prototype.selectOptionsWC;
    /**
     * Initialize component, automatically called once if not blocked by options
     * @return {?}
     */
    NgSelectWebComponent.prototype.initializeWC = function () { };
    /**
     * Initialize options, automaticaly called during init phase, but can be used to reinitialize NgSelectOptions
     * @return {?}
     */
    NgSelectWebComponent.prototype.initOptionsWC = function () { };
    /**
     * Gets instance of plugin by its id
     * @template PluginType
     * @param {?} pluginId Id of plugin, use constants
     * @return {?}
     */
    NgSelectWebComponent.prototype.getPluginWC = function (pluginId) { };
    /**
     * Explicitly runs invalidation of content (change detection)
     * @return {?}
     */
    NgSelectWebComponent.prototype.invalidateVisualsWC = function () { };
    /**
     * Executes actions on NgSelect
     * @param {...?} actions Array of actions that are executed over NgSelect
     * @return {?}
     */
    NgSelectWebComponent.prototype.executeWC = function (actions) { };
    /**
     * Executes function on NgSelect and returns result
     * @template TResult
     * @param {?} func Function that is executed and its result is returned
     * @return {?}
     */
    NgSelectWebComponent.prototype.executeAndReturnWC = function (func) { };
}
