// @ng/common/dist/es2015/src/directives/numberInput/maxValueNumberValidator.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NG_VALIDATORS } from '@angular/forms';
import { Input, forwardRef, Directive, Attribute } from '@angular/core';
import { isBlank, isPresent } from '@asseco/common';
import { Validators } from '../../misc';
/**
 * Validator that is injected with directive MaxValueNumberValidatorDirective
 * @type {?}
 */
const MAX_NUMBER_VALIDATOR = (/** @type {?} */ ({
    provide: NG_VALIDATORS,
    useExisting: forwardRef((/**
     * @return {?}
     */
    () => MaxValueNumberValidatorDirective)),
    multi: true
}));
/**
 * Directive injecting checking for number max value validator
 */
export class MaxValueNumberValidatorDirective {
    //######################### constructor #########################
    /**
     * @param {?} maxValue
     */
    constructor(maxValue) {
        /**
         * Indication whether validator was initialized
         */
        this._initialized = false;
        /** @type {?} */
        var value;
        if (isBlank(maxValue) || maxValue.length < 1 || isNaN(value = parseFloat(maxValue.replace(",", ".")))) {
            value = null;
        }
        this._maxValue = value;
    }
    //######################### public methods - implementation of OnChanges #########################
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (!this._initialized) {
            this._initialized = true;
            this._validator = Validators.max(this._maxValue);
        }
        if ('maxValue' in changes) {
            this._validator = Validators.max(isPresent(this.maxValue) ? (/** @type {?} */ ((/** @type {?} */ (this.maxValue)))) : this._maxValue);
        }
    }
    //######################### public methods - implementation of Validator #########################
    /**
     * Validates input and returns validation result
     * @param {?} control Control that is being validated
     * @return {?} validation results
     */
    validate(control) {
        return this._validator(control);
    }
}
MaxValueNumberValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: "input[number][maxValue][formControlName],input[number][maxValue][formControl],input[number][maxValue][ngModel]",
                providers: [MAX_NUMBER_VALIDATOR]
            },] }
];
/** @nocollapse */
MaxValueNumberValidatorDirective.ctorParameters = () => [
    { type: String, decorators: [{ type: Attribute, args: ["maxValue",] }] }
];
MaxValueNumberValidatorDirective.propDecorators = {
    maxValue: [{ type: Input }]
};
if (false) {
    /**
     * Current max value that is allowed
     * @type {?}
     * @private
     */
    MaxValueNumberValidatorDirective.prototype._maxValue;
    /**
     * Function used for validations
     * @type {?}
     * @private
     */
    MaxValueNumberValidatorDirective.prototype._validator;
    /**
     * Indication whether validator was initialized
     * @type {?}
     * @private
     */
    MaxValueNumberValidatorDirective.prototype._initialized;
    /**
     * Bound max value that is allowed, which overrides value set to attribute maxValue
     * @type {?}
     */
    MaxValueNumberValidatorDirective.prototype.maxValue;
}
