// @ng/common/dist/es2015/src/directives/numberInput/numberInputValidator.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NG_VALIDATORS } from '@angular/forms';
import { forwardRef, Directive } from '@angular/core';
import { Validators } from '../../misc';
/**
 * Validator that is injected with directive NumberInputValidatorDirective
 * @type {?}
 */
const NUMBER_VALIDATOR = (/** @type {?} */ ({
    provide: NG_VALIDATORS,
    useExisting: forwardRef((/**
     * @return {?}
     */
    () => NumberInputValidatorDirective)),
    multi: true
}));
/**
 * Is number directive injecting checking for number validator
 */
export class NumberInputValidatorDirective {
    //######################### public methods - implementation of Validator #########################
    /**
     * Validates input and returns validation result
     * @param {?} control Control that is being validated
     * @return {?} validation results
     */
    validate(control) {
        return Validators.number(control);
    }
}
NumberInputValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: "input[number][formControlName],input[number][formControl],input[number][ngModel]",
                providers: [NUMBER_VALIDATOR]
            },] }
];
