// @ng/common/dist/es2015/src/directives/numberInput/numberInputControlValueAccessor.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, forwardRef, Renderer2, ElementRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { isBlank } from '@asseco/common';
/**
 * Value accessor provider for number inputs
 * @type {?}
 */
const NUMBER_INPUT_VALUE_ACCESSOR = (/** @type {?} */ ({
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef((/**
     * @return {?}
     */
    () => NumberInputControlValueAccessor)),
    multi: true
}));
/**
 * Value accessor for getting and setting values for number inputs
 */
export class NumberInputControlValueAccessor {
    //######################### constructor #########################
    /**
     * @param {?} _renderer
     * @param {?} _elementRef
     */
    constructor(_renderer, _elementRef) {
        this._renderer = _renderer;
        this._elementRef = _elementRef;
        //######################### public properties #########################
        this.onChange = (/**
         * @param {?} _
         * @return {?}
         */
        (_) => { });
        /**
         * Method that is called when picker was touched
         */
        this.onTouched = (/**
         * @return {?}
         */
        () => { });
    }
    //######################### public methods - implementation of ControlValueAccessor #########################
    /**
     * Sets value to select
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this._renderer.setProperty(this._elementRef.nativeElement, 'value', value);
    }
    /**
     * Registers callback that is called when value of select changes
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = (/**
         * @param {?} value
         * @return {?}
         */
        (value) => {
            if (isBlank(value) || value == '') {
                fn(null);
                return;
            }
            if (!/^[+-]?\d+(?:[,.]\d+)?$/g.test(value)) {
                fn(NaN);
                return;
            }
            ;
            fn(parseFloat(value.replace(",", ".")));
        });
    }
    /**
     * Registers callback that is called when select is closed
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
}
NumberInputControlValueAccessor.decorators = [
    { type: Directive, args: [{
                selector: 'input[number][formControlName],input[number][formControl],input[number][ngModel]',
                providers: [NUMBER_INPUT_VALUE_ACCESSOR],
                host: {
                    '(change)': 'onChange($event.target.value)',
                    '(input)': 'onChange($event.target.value)',
                    '(blur)': 'onTouched()'
                }
            },] }
];
/** @nocollapse */
NumberInputControlValueAccessor.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
if (false) {
    /** @type {?} */
    NumberInputControlValueAccessor.prototype.onChange;
    /**
     * Method that is called when picker was touched
     * @type {?}
     */
    NumberInputControlValueAccessor.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    NumberInputControlValueAccessor.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    NumberInputControlValueAccessor.prototype._elementRef;
}
