// @ng/authentication/dist/es5/src/common/userIdentity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Information about user in accessing system
 * @template TUserInfo
 */
var /**
 * Information about user in accessing system
 * @template TUserInfo
 */
UserIdentity = /** @class */ (function () {
    function UserIdentity() {
        //######################### public properties #########################
        //######################### public properties #########################
        /**
         * Indication whether is user authenticated
         */
        this.isAuthenticated = false;
        /**
         * Name of authenticated user
         */
        this.userName = "";
        /**
         * First name of logged user
         */
        this.firstName = "";
        /**
         * Surname of logged user
         */
        this.surname = "";
        /**
         * Array of user permissions
         */
        this.permissions = [];
    }
    return UserIdentity;
}());
/**
 * Information about user in accessing system
 * @template TUserInfo
 */
export { UserIdentity };
if (false) {
    /**
     * Indication whether is user authenticated
     * @type {?}
     */
    UserIdentity.prototype.isAuthenticated;
    /**
     * Name of authenticated user
     * @type {?}
     */
    UserIdentity.prototype.userName;
    /**
     * First name of logged user
     * @type {?}
     */
    UserIdentity.prototype.firstName;
    /**
     * Surname of logged user
     * @type {?}
     */
    UserIdentity.prototype.surname;
    /**
     * Array of user permissions
     * @type {?}
     */
    UserIdentity.prototype.permissions;
    /**
     * Additional information carried with user identity
     * @type {?}
     */
    UserIdentity.prototype.additionalInfo;
}
