// @ng/common/dist/es2015/src/utils/utils.common.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ApplicationRef, InjectionToken } from "@angular/core";
import { enableDebugTools } from '@angular/platform-browser';
import { filter, first } from 'rxjs/operators';
/**
 * Method used for extraction of resolve method for promise
 * @param {?} appStablePromise Promise which contains resolve method that is going to be extracted
 * @return {?}
 */
export function extractAppStableResolve(appStablePromise) {
    return ((/** @type {?} */ (appStablePromise))).__resolve;
}
/**
 * Factory used for creating APP_STABLE promise
 * @return {?}
 */
export function appStablePromiseFactory() {
    /** @type {?} */
    let appStableResolve;
    /** @type {?} */
    let appStablePromise = new Promise((/**
     * @param {?} resolve
     * @return {?}
     */
    resolve => appStableResolve = resolve));
    ((/** @type {?} */ (appStablePromise))).__resolve = appStableResolve;
    return appStablePromise;
}
/**
 * Injection token used for obtaining promise that is resolved when application is first time stable
 * @type {?}
 */
export const APP_STABLE = new InjectionToken("APP_STABLE", { providedIn: 'root', factory: appStablePromiseFactory });
/**
 * Common utility methods
 */
export default class Common {
    //######################### public methods #########################
    /**
     * Runs callback function when angular module is bootstrapped and stable
     * @param {?} moduleRefPromise Promise for module that was bootstrapped
     * @param {?} callback Callback that is called
     * @param {?=} angularProfiler Indication that angular profiler should be enabled
     * @return {?}
     */
    static runWhenModuleStable(moduleRefPromise, callback, angularProfiler) {
        angularProfiler = angularProfiler || false;
        moduleRefPromise.then((/**
         * @param {?} moduleRef
         * @return {?}
         */
        (moduleRef) => {
            /** @type {?} */
            const appRef = moduleRef.injector.get(ApplicationRef);
            appRef.isStable
                .pipe(filter((/**
             * @param {?} isStable
             * @return {?}
             */
            (isStable) => isStable)), first())
                .subscribe((/**
             * @return {?}
             */
            () => {
                /** @type {?} */
                let appStablePromise = moduleRef.injector.get(APP_STABLE);
                if (angularProfiler) {
                    enableDebugTools(appRef.components[0]);
                }
                callback(moduleRef);
                if (appStablePromise) {
                    /** @type {?} */
                    let resolveAsStable = extractAppStableResolve(appStablePromise);
                    resolveAsStable();
                }
            }));
        }));
    }
}
