// @ng/sample/dist/es5/src/modules/ngSelect.module.js
import * as ɵngcc0 from '@angular/core';
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgSelectValuePipe } from '../pipes/ngSelectValue.pipe';
import { OptGroupComponent } from '../components/option/optgroup.component';
import { OptionComponent } from '../components/option/option.component';
import { NgCommonComponent } from '../components/ngCommon/ngCommon.component';
/**
 * Module for select and its options
 */
var NgSelectModule = /** @class */ (function () {
    function NgSelectModule() {
    }
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(NgSelectModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    OptionComponent,
                    OptGroupComponent,
                    NgSelectValuePipe,
                    NgCommonComponent
                ],
                exports: [
                    OptionComponent,
                    OptGroupComponent,
                    NgSelectValuePipe,
                    NgCommonComponent
                ]
            }]
    }], function () { return []; }, null);
NgSelectModule.ngModuleDef = ɵngcc0.ɵdefineNgModule({ type: NgSelectModule, declarations: function () { return [OptionComponent,
        OptGroupComponent,
        NgSelectValuePipe,
        NgCommonComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [OptionComponent,
        OptGroupComponent,
        NgSelectValuePipe,
        NgCommonComponent]; } });
NgSelectModule.ngInjectorDef = ɵngcc0.defineInjector({ factory: function NgSelectModule_Factory(t) { return new (t || NgSelectModule)(); }, providers: [], imports: [[
            CommonModule
        ],
        [
            OptionComponent,
            OptGroupComponent,
            NgSelectValuePipe,
            NgCommonComponent
        ]] });
    return NgSelectModule;
}());
export { NgSelectModule };
