// @ng/sample/dist/es5/src/modules/ngSelect.module.js
import * as ɵngcc0 from '@angular/core';
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgSelectValuePipe } from '../pipes/ngSelectValue.pipe';
import { OptGroupComponent } from '../components/option/optgroup.component';
import { OptionComponent } from '../components/option/option.component';
/**
 * Module for select and its options
 */
var NgSelectModule = /** @class */ (function () {
    function NgSelectModule() {
    }
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(NgSelectModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    OptionComponent,
                    OptGroupComponent,
                    NgSelectValuePipe
                ],
                exports: [
                    OptionComponent,
                    OptGroupComponent,
                    NgSelectValuePipe
                ]
            }]
    }], function () { return []; }, null);
NgSelectModule.ngModuleDef = ɵngcc0.ɵdefineNgModule({ type: NgSelectModule, declarations: [OptionComponent,
        OptGroupComponent,
        NgSelectValuePipe], imports: [CommonModule], exports: [OptionComponent,
        OptGroupComponent,
        NgSelectValuePipe] });
NgSelectModule.ngInjectorDef = ɵngcc0.defineInjector({ factory: function NgSelectModule_Factory(t) { return new (t || NgSelectModule)(); }, providers: [], imports: [[
            CommonModule
        ],
        [
            OptionComponent,
            OptGroupComponent,
            NgSelectValuePipe
        ]] });
    return NgSelectModule;
}());
export { NgSelectModule };
