// @ng/select/dist/es5/src/plugins/valueHandler/dynamic/dynamicValueHandler.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Inject, Optional, ElementRef } from '@angular/core';
import { extend, isBlank } from '@asseco/common';
import { NG_SELECT_PLUGIN_INSTANCES } from '../../../components/select/types';
import { VALUE_HANDLER_OPTIONS } from '../types';
import { ValueHandlerBase } from '../valueHandlerBase';
var ɵ0 = /**
 * @param {?} value
 * @return {?}
 */
function (value) { return value; };
/**
 * Default options for value handler
 * \@internal
 * @type {?}
 */
var defaultOptions = {
    textExtractor: (ɵ0)
};
/**
 * Component used for handling current value of NgSelect, allows values which are not present in options
 * @template TValue
 */
var DynamicValueHandlerComponent = /** @class */ (function (_super) {
    __extends(DynamicValueHandlerComponent, _super);
    //######################### constructor #########################
    function DynamicValueHandlerComponent(ngSelectPlugins, pluginElement, options) {
        var _this = _super.call(this, ngSelectPlugins, pluginElement) || this;
        //######################### protected methods #########################
        /**
         * Sets value
         */
        _this._setValue = (/**
         * @param {?} option
         * @return {?}
         */
        function (option) {
            //multiple values are allowed
            if (_this.options.multiple) {
                if (!Array.isArray(_this.selectedOptions)) {
                    _this.selectedOptions = [];
                }
                else {
                    /** @type {?} */
                    var opt_1;
                    //value exists, removing from list
                    if ((opt_1 = _this.selectedOptions.find((/**
                     * @param {?} selOpt
                     * @return {?}
                     */
                    function (selOpt) { return _this.valueComparer(selOpt.value, opt_1.value); })))) {
                        /** @type {?} */
                        var index = _this.selectedOptions.indexOf(option);
                        _this.selectedOptions.splice(index, 1);
                    }
                    //adding value
                    else {
                        _this.selectedOptions.push(option);
                    }
                }
            }
            else 
            //only signle value allowed
            {
                _this.selectedOptions = option;
            }
            _this._clearSelected();
            _this._markValueAsSelected();
            _this._normalState.invalidateVisuals();
            _this.valueChange.emit();
            //close popup if not multiple
            if (!_this.options.multiple) {
                _this.popupVisibilityRequest.emit(false);
            }
            else {
                _this._popup.invalidateVisuals();
            }
        });
        _this._options = extend(true, {}, defaultOptions, options);
        return _this;
    }
    Object.defineProperty(DynamicValueHandlerComponent.prototype, "options", {
        //######################### public properties - implementation of DynamicValueHandler #########################
        /**
         * Options for NgSelect plugin
         */
        get: 
        //######################### public properties - implementation of DynamicValueHandler #########################
        /**
         * Options for NgSelect plugin
         * @return {?}
         */
        function () {
            return this._options;
        },
        set: /**
         * @param {?} options
         * @return {?}
         */
        function (options) {
            this._options = extend(true, this._options, options);
        },
        enumerable: true,
        configurable: true
    });
    //######################### public methods - implementation of DynamicValueHandler #########################
    /**
     * Sets value for NgSelect
     * @param value Value to be set
     */
    //######################### public methods - implementation of DynamicValueHandler #########################
    /**
     * Sets value for NgSelect
     * @param {?} value Value to be set
     * @return {?}
     */
    DynamicValueHandlerComponent.prototype.setValue = 
    //######################### public methods - implementation of DynamicValueHandler #########################
    /**
     * Sets value for NgSelect
     * @param {?} value Value to be set
     * @return {?}
     */
    function (value) {
        this._useOptionsAsValue(value);
    };
    /**
     * Initialize plugin options, all operations required to be done with plugin options are handled here
     */
    /**
     * Initialize plugin options, all operations required to be done with plugin options are handled here
     * @return {?}
     */
    DynamicValueHandlerComponent.prototype.initOptions = /**
     * Initialize plugin options, all operations required to be done with plugin options are handled here
     * @return {?}
     */
    function () {
    };
    /**
     * Explicitly runs invalidation of content (change detection)
     */
    /**
     * Explicitly runs invalidation of content (change detection)
     * @return {?}
     */
    DynamicValueHandlerComponent.prototype.invalidateVisuals = /**
     * Explicitly runs invalidation of content (change detection)
     * @return {?}
     */
    function () {
    };
    /**
     * Loads options
     */
    /**
     * Loads options
     * @protected
     * @return {?}
     */
    DynamicValueHandlerComponent.prototype._loadOptions = /**
     * Loads options
     * @protected
     * @return {?}
     */
    function () {
    };
    /**
     * Converts value to options
     * @param value Value to be changed to options
     */
    /**
     * Converts value to options
     * @protected
     * @param {?} value Value to be changed to options
     * @return {?}
     */
    DynamicValueHandlerComponent.prototype._useOptionsAsValue = /**
     * Converts value to options
     * @protected
     * @param {?} value Value to be changed to options
     * @return {?}
     */
    function (value) {
        return __awaiter(this, void 0, void 0, function () {
            var items, options, _i, items_1, itm, _a, _b, item, _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        //set empty value
                        if (isBlank(value) || (Array.isArray(value) && !value.length)) {
                            this.selectedOptions = value;
                            this._clearSelected();
                            this._normalState.invalidateVisuals();
                            this.valueChange.emit();
                            return [2 /*return*/];
                        }
                        if (!this.options.multiple) return [3 /*break*/, 7];
                        if (!Array.isArray(value)) return [3 /*break*/, 5];
                        items = value;
                        options = [];
                        _i = 0, items_1 = items;
                        _d.label = 1;
                    case 1:
                        if (!(_i < items_1.length)) return [3 /*break*/, 4];
                        itm = items_1[_i];
                        _b = (_a = options).push;
                        return [4 /*yield*/, this._loadText(itm)];
                    case 2:
                        _b.apply(_a, [_d.sent()]);
                        _d.label = 3;
                    case 3:
                        _i++;
                        return [3 /*break*/, 1];
                    case 4:
                        this.selectedOptions = options;
                        return [3 /*break*/, 6];
                    case 5: throw new Error('Don`t you have redundant "multiple"?');
                    case 6: return [3 /*break*/, 10];
                    case 7:
                        if (!Array.isArray(value)) return [3 /*break*/, 8];
                        throw new Error('Are you missing attribute "multiple"?');
                    case 8:
                        item = value;
                        _c = this;
                        return [4 /*yield*/, this._loadText(item)];
                    case 9:
                        _c.selectedOptions = _d.sent();
                        _d.label = 10;
                    case 10:
                        this._clearSelected();
                        this._markValueAsSelected();
                        this._normalState.invalidateVisuals();
                        this.valueChange.emit();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Loads text for specified value
     * @param value Value that is going to be used for obtaining option
     */
    /**
     * Loads text for specified value
     * @protected
     * @param {?} value Value that is going to be used for obtaining option
     * @return {?}
     */
    DynamicValueHandlerComponent.prototype._loadText = /**
     * Loads text for specified value
     * @protected
     * @param {?} value Value that is going to be used for obtaining option
     * @return {?}
     */
    function (value) {
        return __awaiter(this, void 0, void 0, function () {
            var opts, opt;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.options.dynamicOptionsCallback) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.options.dynamicOptionsCallback(value)];
                    case 1:
                        opts = _a.sent();
                        if (opts && opts.length) {
                            opt = opts[0];
                            opt.value = value;
                            opt.selected = true;
                            return [2 /*return*/, opt];
                        }
                        _a.label = 2;
                    case 2: 
                    //load option from value
                    return [2 /*return*/, (/** @type {?} */ ({
                            selected: true,
                            active: false,
                            value: value,
                            text: this.options.textExtractor(value)
                        }))];
                }
            });
        });
    };
    DynamicValueHandlerComponent.decorators = [
        { type: Component, args: [{
                    selector: "ng-dynamic-value-handler",
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    DynamicValueHandlerComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [NG_SELECT_PLUGIN_INSTANCES,] }, { type: Optional }] },
        { type: ElementRef },
        { type: undefined, decorators: [{ type: Inject, args: [VALUE_HANDLER_OPTIONS,] }, { type: Optional }] }
    ]; };
    return DynamicValueHandlerComponent;
}(ValueHandlerBase));
export { DynamicValueHandlerComponent };
if (false) {
    /**
     * Sets value
     * @type {?}
     * @protected
     */
    DynamicValueHandlerComponent.prototype._setValue;
}
export { ɵ0 };
