// @ng/select/dist/es5/src/components/select/select.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Interface describing object storing all existing plugin instances for NgSelect
 * @record
 */
export function NgSelectPluginInstances() { }
/**
 * Public API for NgSelect
 * @record
 * @template TValue
 */
export function NgSelect() { }
if (false) {
    /**
     * Occurs every time when NgSelect is initialized or reinitialized, if value is false NgSelect was not initialized yet
     * @type {?}
     */
    NgSelect.prototype.initialized;
    /**
     * Gets current state of initialization
     * @type {?}
     */
    NgSelect.prototype.isInitialized;
    /**
     * Gets or sets NgSelect options
     * @type {?}
     */
    NgSelect.prototype.selectOptions;
    /**
     * Initialize component, automatically called once if not blocked by options
     * @return {?}
     */
    NgSelect.prototype.initialize = function () { };
    /**
     * Initialize options, automaticaly called during init phase, but can be used to reinitialize NgSelectOptions
     * @return {?}
     */
    NgSelect.prototype.initOptions = function () { };
    /**
     * Gets instance of plugin by its id
     * @template PluginType
     * @param {?} pluginId Id of plugin, use constants
     * @return {?}
     */
    NgSelect.prototype.getPlugin = function (pluginId) { };
    /**
     * Explicitly runs invalidation of content (change detection)
     * @return {?}
     */
    NgSelect.prototype.invalidateVisuals = function () { };
    /**
     * Executes actions on NgSelect
     * @param {...?} actions Array of actions that are executed over NgSelect
     * @return {?}
     */
    NgSelect.prototype.execute = function (actions) { };
    /**
     * Executes function on NgSelect and returns result
     * @template TResult
     * @param {?} func Function that is executed and its result is returned
     * @return {?}
     */
    NgSelect.prototype.executeAndReturn = function (func) { };
}
