// @ng/common/dist/es5/src/components/progressIndicator/progressIndicator.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Optional, Injectable, PLATFORM_ID, Inject } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { ProgressIndicatorOptions } from './progressIndicatorOptions';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "./progressIndicatorOptions";
/**
 * Service that is used for displaying and hiding progress indicator
 */
var ProgressIndicatorService = /** @class */ (function () {
    //######################### constructors #########################
    function ProgressIndicatorService(platformId, config) {
        this.config = config;
        //######################### private fields #########################
        /**
         * Indication that current code is running in browser
         */
        this._isBrowser = false;
        /**
         * Number of running requests
         */
        this._runningRequests = 0;
        /**
         * Used for invoking event runningChanged
         */
        this._runningChanged = new Subject();
        //######################### public properties #########################
        /**
         * Indication that progress indicator is running
         */
        this.running = false;
        this._isBrowser = isPlatformBrowser(platformId);
        if (config && !(config instanceof ProgressIndicatorOptions)) {
            this.config = undefined;
            console.warn("Provided configuration for 'ProgressIndicatorService' is not of type 'ProgressIndicatorOptions' and will be ignored!");
        }
        this.config = config || new ProgressIndicatorOptions();
    }
    Object.defineProperty(ProgressIndicatorService.prototype, "runningChanged", {
        /**
         * Occurs when 'running' changes
         */
        get: /**
         * Occurs when 'running' changes
         * @return {?}
         */
        function () {
            return this._runningChanged.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    //######################### public methods #########################
    /**
     * Displays progress indicator after short delay
     */
    //######################### public methods #########################
    /**
     * Displays progress indicator after short delay
     * @return {?}
     */
    ProgressIndicatorService.prototype.showProgress = 
    //######################### public methods #########################
    /**
     * Displays progress indicator after short delay
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this._isBrowser) {
            return;
        }
        if (!this._timeout && this._runningRequests < 1) {
            this._timeout = setTimeout((/**
             * @return {?}
             */
            function () {
                _this._onRunning(true);
                clearTimeout(_this._timeout);
                _this._timeout = null;
            }), this.config.timeout);
        }
        this._runningRequests++;
    };
    /**
     * Hides progress indicator
     * @param force Indication that indicator should be hidden even if some processes are running
     */
    /**
     * Hides progress indicator
     * @param {?=} force Indication that indicator should be hidden even if some processes are running
     * @return {?}
     */
    ProgressIndicatorService.prototype.hideProgress = /**
     * Hides progress indicator
     * @param {?=} force Indication that indicator should be hidden even if some processes are running
     * @return {?}
     */
    function (force) {
        if (!this._isBrowser) {
            return;
        }
        if (force) {
            this._runningRequests = 0;
        }
        if (this._runningRequests > 0) {
            this._runningRequests--;
        }
        if (this._runningRequests < 1) {
            clearTimeout(this._timeout);
            this._timeout = null;
            this._onRunning(false);
        }
    };
    //######################### private methods #########################
    /**
     * Used for invoking 'runningChanged' event
     */
    //######################### private methods #########################
    /**
     * Used for invoking 'runningChanged' event
     * @private
     * @param {?} value
     * @return {?}
     */
    ProgressIndicatorService.prototype._onRunning = 
    //######################### private methods #########################
    /**
     * Used for invoking 'runningChanged' event
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.running = value;
        this._runningChanged.next(value);
    };
    ProgressIndicatorService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    ProgressIndicatorService.ctorParameters = function () { return [
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
        { type: ProgressIndicatorOptions, decorators: [{ type: Optional }] }
    ]; };
    /** @nocollapse */ ProgressIndicatorService.ngInjectableDef = i0.defineInjectable({ factory: function ProgressIndicatorService_Factory() { return new ProgressIndicatorService(i0.inject(i0.PLATFORM_ID), i0.inject(i1.ProgressIndicatorOptions, 8)); }, token: ProgressIndicatorService, providedIn: "root" });
    return ProgressIndicatorService;
}());
export { ProgressIndicatorService };
if (false) {
    /**
     * Indication that current code is running in browser
     * @type {?}
     * @private
     */
    ProgressIndicatorService.prototype._isBrowser;
    /**
     * Id of running timeout
     * @type {?}
     * @private
     */
    ProgressIndicatorService.prototype._timeout;
    /**
     * Number of running requests
     * @type {?}
     * @private
     */
    ProgressIndicatorService.prototype._runningRequests;
    /**
     * Used for invoking event runningChanged
     * @type {?}
     * @private
     */
    ProgressIndicatorService.prototype._runningChanged;
    /**
     * Indication that progress indicator is running
     * @type {?}
     */
    ProgressIndicatorService.prototype.running;
    /** @type {?} */
    ProgressIndicatorService.prototype.config;
}
