// @ng/common/dist/es5/src/components/progressIndicator/progressIndicator.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ProgressIndicatorService } from './progressIndicator.service';
/**
 * Component that is used for displaying progress indicator
 */
var ProgressIndicatorComponent = /** @class */ (function () {
    //######################### constructor #########################
    function ProgressIndicatorComponent(_service) {
        var _this = this;
        this._service = _service;
        //######################### public properties - inputs #########################
        /**
         * Applied css classes
         */
        this.appliedClass = { "progress-indicator": true };
        this.running = this._service.running;
        this._subscription = this._service.runningChanged.subscribe((/**
         * @param {?} running
         * @return {?}
         */
        function (running) { return _this.running = running; }));
    }
    Object.defineProperty(ProgressIndicatorComponent.prototype, "cssClass", {
        /**
         * Sets css classes that will be applied to indicator
         */
        set: /**
         * Sets css classes that will be applied to indicator
         * @param {?} cssClass
         * @return {?}
         */
        function (cssClass) {
            this.appliedClass = {};
            this.appliedClass[cssClass] = true;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressIndicatorComponent.prototype, "running", {
        //######################### private properties #########################
        /**
         * Sets indication that progress indicator is running
         */
        set: 
        //######################### private properties #########################
        /**
         * Sets indication that progress indicator is running
         * @private
         * @param {?} running
         * @return {?}
         */
        function (running) {
            this.appliedClass["running"] = running;
        },
        enumerable: true,
        configurable: true
    });
    ;
    //######################### public methods - implementation of OnDestroy #########################
    /**
     * Called when component is destroyed
     */
    //######################### public methods - implementation of OnDestroy #########################
    /**
     * Called when component is destroyed
     * @return {?}
     */
    ProgressIndicatorComponent.prototype.ngOnDestroy = 
    //######################### public methods - implementation of OnDestroy #########################
    /**
     * Called when component is destroyed
     * @return {?}
     */
    function () {
        if (this._subscription) {
            this._subscription.unsubscribe();
            this._subscription = null;
        }
    };
    ProgressIndicatorComponent.decorators = [
        { type: Component, args: [{
                    selector: "progress-indicator",
                    template: "<div [ngClass]=\"appliedClass\">\r\n    <div></div>\r\n</div>",
                    styles: ["\r\n.progress-indicator\r\n{\r\n    -webkit-transition: all 250ms linear;\r\n    -moz-transition: all 250ms linear;\r\n    transition: all 250ms linear;\r\n    background-color: rgba(0, 0, 0, 0.3);\r\n    height: 100%;\r\n    width: 100%;\r\n    position: fixed;\r\n    top: 0;\r\n    left: 0;\r\n    z-index: 23456;\r\n    opacity: 0;\r\n    visibility: hidden;\r\n}\r\n.progress-indicator.running\r\n{\r\n    visibility: visible;\r\n    opacity: 1;\r\n}\r\n.progress-indicator.running > div\r\n{\r\n    border-left: 4px solid #eee;\r\n    border-radius: 25px;\r\n    border-top: 4px solid #eee;\r\n    height: 40px;\r\n    margin-left: auto;\r\n    margin-right: auto;\r\n    position: relative;\r\n    top: 50%;\r\n    width: 40px;\r\n    animation-name: rotate-progress;\r\n    animation-duration: 350ms;\r\n    animation-iteration-count: infinite;\r\n    animation-timing-function: linear;\r\n}\r\n\r\n@keyframes rotate-progress\r\n{\r\n    from\r\n    {\r\n        transform: rotate(0deg);\r\n    }\r\n    to\r\n    {\r\n        transform: rotate(360deg);\r\n    }\r\n}"]
                }] }
    ];
    /** @nocollapse */
    ProgressIndicatorComponent.ctorParameters = function () { return [
        { type: ProgressIndicatorService }
    ]; };
    ProgressIndicatorComponent.propDecorators = {
        cssClass: [{ type: Input }]
    };
    return ProgressIndicatorComponent;
}());
export { ProgressIndicatorComponent };
if (false) {
    /**
     * Subscription for changes in ProgressIndicatorService
     * @type {?}
     * @private
     */
    ProgressIndicatorComponent.prototype._subscription;
    /**
     * Applied css classes
     * @type {?}
     */
    ProgressIndicatorComponent.prototype.appliedClass;
    /**
     * @type {?}
     * @private
     */
    ProgressIndicatorComponent.prototype._service;
    /* Skipping unhandled member: ;*/
}
