// @ng/common/dist/es5/src/utils/utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import RouterHelper from './utils.routerHelper';
import Common from './utils.common';
import Forms from './utils.forms';
/**
 * Helper methods for angular2 router
 * @record
 */
export function IRouterHelper() { }
if (false) {
    /**
     * Extracts route definitions from components if routes are set using decorator ComponentRoute
     * @param {?} components Array of components to be used for extraction
     * @return {?} RouteDefinition Extracted routes
     */
    IRouterHelper.prototype.extractRoutes = function (components) { };
}
/**
 * Common utility methods
 * @record
 */
export function ICommon() { }
if (false) {
    /**
     * Runs callback function when angular module is bootstrapped and stable
     * @param {?} moduleRefPromise Promise for module that was bootstrapped
     * @param {?} callback Callback that is called
     * @param {?=} angularProfiler Indication that angular profiler should be enabled
     * @return {?}
     */
    ICommon.prototype.runWhenModuleStable = function (moduleRefPromise, callback, angularProfiler) { };
}
/**
 * Utility methods form angular 2 forms
 * @record
 */
export function IForms() { }
if (false) {
    /**
     * Gets indication whether controls have errors, with custom indication of submitted
     * @param {?} form Form containing controls
     * @param {?} controls Array of controls names to be checked for errors
     * @param {?=} submitted Indication whether form was submitted, defaults to false
     * @return {?}
     */
    IForms.prototype.hasErrorCustom = function (form, controls, submitted) { };
    /**
     * Gets indication whether hide alerts or not for control, with custom indication of submitted
     * @param {?} form Form containing controls
     * @param {?} control Controls name that will be checked
     * @param {?=} errors Array of validation errors to be checked for existance
     * @param {?=} submitted Indication whether form was submitted, defaults to false
     * @return {?}
     */
    IForms.prototype.alertHiddenCustom = function (form, control, errors, submitted) { };
    /**
     * Gets indication whether controls have errors
     * @param {?} form Form containing controls
     * @param {?} controls Array of controls names to be checked for errors
     * @return {?}
     */
    IForms.prototype.hasError = function (form, controls) { };
    /**
     * Gets indication whether hide alerts or not for control
     * @param {?} form Form containing controls
     * @param {?} control Controls name that will be checked
     * @param {?} errors Array of validation errors to be checked for existance
     * @return {?}
     */
    IForms.prototype.alertHidden = function (form, control, errors) { };
}
/**
 * Utils methods categories
 */
var /**
 * Utils methods categories
 */
Utils = /** @class */ (function () {
    function Utils() {
    }
    Object.defineProperty(Utils, "routerHelper", {
        /**
         * Angular 2 router helper
         */
        get: /**
         * Angular 2 router helper
         * @return {?}
         */
        function () {
            return RouterHelper;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Utils, "common", {
        /**
         * Common utility methods
         */
        get: /**
         * Common utility methods
         * @return {?}
         */
        function () {
            return Common;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Utils, "forms", {
        /**
         * Utility methods form angular 2 forms
         */
        get: /**
         * Utility methods form angular 2 forms
         * @return {?}
         */
        function () {
            return Forms;
        },
        enumerable: true,
        configurable: true
    });
    return Utils;
}());
export { Utils };
