// @ng/common/dist/es5/src/directives/numberInput/minValueNumberValidator.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NG_VALIDATORS } from '@angular/forms';
import { forwardRef, Directive, Attribute, Input } from '@angular/core';
import { isBlank, isPresent } from '@asseco/common';
import { Validators } from '../../misc';
/**
 * Validator that is injected with directive MinValueNumberValidatorDirective
 * @type {?}
 */
var MIN_NUMBER_VALIDATOR = (/** @type {?} */ ({
    provide: NG_VALIDATORS,
    useExisting: forwardRef((/**
     * @return {?}
     */
    function () { return MinValueNumberValidatorDirective; })),
    multi: true
}));
/**
 * Directive injecting checking for number min value validator
 */
var MinValueNumberValidatorDirective = /** @class */ (function () {
    //######################### constructor #########################
    function MinValueNumberValidatorDirective(minValue) {
        /**
         * Indication whether validator was initialized
         */
        this._initialized = false;
        /** @type {?} */
        var value;
        if (isBlank(minValue) || minValue.length < 1 || isNaN(value = parseFloat(minValue.replace(",", ".")))) {
            value = null;
        }
        this._minValue = value;
    }
    //######################### public methods - implementation of OnChanges #########################
    //######################### public methods - implementation of OnChanges #########################
    /**
     * @param {?} changes
     * @return {?}
     */
    MinValueNumberValidatorDirective.prototype.ngOnChanges = 
    //######################### public methods - implementation of OnChanges #########################
    /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (!this._initialized) {
            this._initialized = true;
            this._validator = Validators.min(this._minValue);
        }
        if ('minValue' in changes) {
            this._validator = Validators.min(isPresent(this.minValue) ? (/** @type {?} */ ((/** @type {?} */ (this.minValue)))) : this._minValue);
        }
    };
    //######################### public methods - implementation of Validator #########################
    /**
     * Validates input and returns validation result
     * @param control Control that is being validated
     * @returns validation results
     */
    //######################### public methods - implementation of Validator #########################
    /**
     * Validates input and returns validation result
     * @param {?} control Control that is being validated
     * @return {?} validation results
     */
    MinValueNumberValidatorDirective.prototype.validate = 
    //######################### public methods - implementation of Validator #########################
    /**
     * Validates input and returns validation result
     * @param {?} control Control that is being validated
     * @return {?} validation results
     */
    function (control) {
        return this._validator(control);
    };
    MinValueNumberValidatorDirective.decorators = [
        { type: Directive, args: [{
                    selector: "input[number][minValue][formControlName],input[number][minValue][formControl],input[number][minValue][ngModel]",
                    providers: [MIN_NUMBER_VALIDATOR]
                },] }
    ];
    /** @nocollapse */
    MinValueNumberValidatorDirective.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Attribute, args: ["minValue",] }] }
    ]; };
    MinValueNumberValidatorDirective.propDecorators = {
        minValue: [{ type: Input }]
    };
    return MinValueNumberValidatorDirective;
}());
export { MinValueNumberValidatorDirective };
if (false) {
    /**
     * Current min value that is allowed
     * @type {?}
     * @private
     */
    MinValueNumberValidatorDirective.prototype._minValue;
    /**
     * Function used for validations
     * @type {?}
     * @private
     */
    MinValueNumberValidatorDirective.prototype._validator;
    /**
     * Indication whether validator was initialized
     * @type {?}
     * @private
     */
    MinValueNumberValidatorDirective.prototype._initialized;
    /**
     * Bound min value that is allowed, which overrides value set to attribute minValue
     * @type {?}
     */
    MinValueNumberValidatorDirective.prototype.minValue;
}
