// @ng/common/dist/es5/src/directives/numberInput/numberInputValidator.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NG_VALIDATORS } from '@angular/forms';
import { forwardRef, Directive } from '@angular/core';
import { Validators } from '../../misc';
/**
 * Validator that is injected with directive NumberInputValidatorDirective
 * @type {?}
 */
var NUMBER_VALIDATOR = (/** @type {?} */ ({
    provide: NG_VALIDATORS,
    useExisting: forwardRef((/**
     * @return {?}
     */
    function () { return NumberInputValidatorDirective; })),
    multi: true
}));
/**
 * Is number directive injecting checking for number validator
 */
var NumberInputValidatorDirective = /** @class */ (function () {
    function NumberInputValidatorDirective() {
    }
    //######################### public methods - implementation of Validator #########################
    /**
     * Validates input and returns validation result
     * @param control Control that is being validated
     * @returns validation results
     */
    //######################### public methods - implementation of Validator #########################
    /**
     * Validates input and returns validation result
     * @param {?} control Control that is being validated
     * @return {?} validation results
     */
    NumberInputValidatorDirective.prototype.validate = 
    //######################### public methods - implementation of Validator #########################
    /**
     * Validates input and returns validation result
     * @param {?} control Control that is being validated
     * @return {?} validation results
     */
    function (control) {
        return Validators.number(control);
    };
    NumberInputValidatorDirective.decorators = [
        { type: Directive, args: [{
                    selector: "input[number][formControlName],input[number][formControl],input[number][ngModel]",
                    providers: [NUMBER_VALIDATOR]
                },] }
    ];
    return NumberInputValidatorDirective;
}());
export { NumberInputValidatorDirective };
