// @ng/common/dist/es5/src/directives/numberInput/numberInputControlValueAccessor.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, forwardRef, Renderer2, ElementRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { isBlank } from '@asseco/common';
/**
 * Value accessor provider for number inputs
 * @type {?}
 */
var NUMBER_INPUT_VALUE_ACCESSOR = (/** @type {?} */ ({
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef((/**
     * @return {?}
     */
    function () { return NumberInputControlValueAccessor; })),
    multi: true
}));
/**
 * Value accessor for getting and setting values for number inputs
 */
var NumberInputControlValueAccessor = /** @class */ (function () {
    //######################### constructor #########################
    function NumberInputControlValueAccessor(_renderer, _elementRef) {
        this._renderer = _renderer;
        this._elementRef = _elementRef;
        //######################### public properties #########################
        this.onChange = (/**
         * @param {?} _
         * @return {?}
         */
        function (_) { });
        /**
         * Method that is called when picker was touched
         */
        this.onTouched = (/**
         * @return {?}
         */
        function () { });
    }
    //######################### public methods - implementation of ControlValueAccessor #########################
    /**
     * Sets value to select
     */
    //######################### public methods - implementation of ControlValueAccessor #########################
    /**
     * Sets value to select
     * @param {?} value
     * @return {?}
     */
    NumberInputControlValueAccessor.prototype.writeValue = 
    //######################### public methods - implementation of ControlValueAccessor #########################
    /**
     * Sets value to select
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this._renderer.setProperty(this._elementRef.nativeElement, 'value', value);
    };
    /**
     * Registers callback that is called when value of select changes
     */
    /**
     * Registers callback that is called when value of select changes
     * @param {?} fn
     * @return {?}
     */
    NumberInputControlValueAccessor.prototype.registerOnChange = /**
     * Registers callback that is called when value of select changes
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = (/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (isBlank(value) || value == '') {
                fn(null);
                return;
            }
            if (!/^[+-]?\d+(?:[,.]\d+)?$/g.test(value)) {
                fn(NaN);
                return;
            }
            ;
            fn(parseFloat(value.replace(",", ".")));
        });
    };
    /**
     * Registers callback that is called when select is closed
     */
    /**
     * Registers callback that is called when select is closed
     * @param {?} fn
     * @return {?}
     */
    NumberInputControlValueAccessor.prototype.registerOnTouched = /**
     * Registers callback that is called when select is closed
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    NumberInputControlValueAccessor.decorators = [
        { type: Directive, args: [{
                    selector: 'input[number][formControlName],input[number][formControl],input[number][ngModel]',
                    providers: [NUMBER_INPUT_VALUE_ACCESSOR],
                    host: {
                        '(change)': 'onChange($event.target.value)',
                        '(input)': 'onChange($event.target.value)',
                        '(blur)': 'onTouched()'
                    }
                },] }
    ];
    /** @nocollapse */
    NumberInputControlValueAccessor.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    return NumberInputControlValueAccessor;
}());
export { NumberInputControlValueAccessor };
if (false) {
    /** @type {?} */
    NumberInputControlValueAccessor.prototype.onChange;
    /**
     * Method that is called when picker was touched
     * @type {?}
     */
    NumberInputControlValueAccessor.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    NumberInputControlValueAccessor.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    NumberInputControlValueAccessor.prototype._elementRef;
}
