// @ng/select/dist/es5/src/misc/types.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from "@angular/core";
/**
 * Injection token for 'NgSelectOptions'
 * @type {?}
 */
export var NG_SELECT_OPTIONS = new InjectionToken('NG_SELECT_OPTIONS');
/**
 * Injection token for 'KeyboardHandler' implementation
 * @type {?}
 */
export var KEYBOARD_HANDLER_TYPE = new InjectionToken('KEYBOARD_HANDLER_TYPE');
/**
 * Injection token for 'NormalState' implementation
 * @type {?}
 */
export var NORMAL_STATE_TYPE = new InjectionToken('NORMAL_STATE_TYPE');
/**
 * Injection token for 'Popup' implementation
 * @type {?}
 */
export var POPUP_TYPE = new InjectionToken('POPUP_TYPE');
/**
 * Injection token for 'Positioner' implementation
 * @type {?}
 */
export var POSITIONER_TYPE = new InjectionToken('POSITIONER_TYPE');
/**
 * Injection token for 'ReadonlyState' implementation
 * @type {?}
 */
export var READONLY_STATE_TYPE = new InjectionToken('READONLY_STATE_TYPE');
/**
 * Injection token for 'ValueHandler' implementation
 * @type {?}
 */
export var VALUE_HANDLER_TYPE = new InjectionToken('VALUE_HANDLER_TYPE');
/**
 * Injection token for 'LiveSearch' implementation
 * @type {?}
 */
export var LIVE_SEARCH_TYPE = new InjectionToken('LIVE_SEARCH_TYPE');
/**
 * Injection token for 'TextsLocator' implementation
 * @type {?}
 */
export var TEXTS_LOCATOR_TYPE = new InjectionToken('TEXTS_LOCATOR_TYPE');
