// @ng/select/dist/es2015/src/plugins/liveSearch/no/noLiveSearch.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Inject, Optional, ElementRef, ViewChild, EventEmitter } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { NG_SELECT_PLUGIN_INSTANCES } from '../../../components/select/types';
/**
 * Component used for no live search
 */
export class NoLiveSearchComponent {
    //######################### constructor #########################
    /**
     * @param {?} ngSelectPlugins
     * @param {?} pluginElement
     * @param {?} _document
     */
    constructor(ngSelectPlugins, pluginElement, _document) {
        this.ngSelectPlugins = ngSelectPlugins;
        this.pluginElement = pluginElement;
        this._document = _document;
        /**
         * Current value of live search
         */
        this.searchValue = null;
        /**
         * Occurs when current value of live search changes
         */
        this.searchValueChange = new EventEmitter();
    }
    /**
     * HTML element that represents live search
     * @return {?}
     */
    get liveSearchElement() {
        return this._document.createElement("span");
    }
    //######################### public methods - implementation of NoLiveSearch #########################
    /**
     * Initialize plugin, to be ready to use, initialize communication with other plugins
     * @return {?}
     */
    initialize() {
    }
    /**
     * Initialize plugin options, all operations required to be done with plugin options are handled here
     * @return {?}
     */
    initOptions() {
    }
    /**
     * Explicitly runs invalidation of content (change detection)
     * @return {?}
     */
    invalidateVisuals() {
    }
}
NoLiveSearchComponent.decorators = [
    { type: Component, args: [{
                selector: "ng-no-live-search",
                template: '',
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
NoLiveSearchComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [NG_SELECT_PLUGIN_INSTANCES,] }, { type: Optional }] },
    { type: ElementRef },
    { type: HTMLDocument, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
NoLiveSearchComponent.propDecorators = {
    liveSearchElementChild: [{ type: ViewChild, args: ['liveSearchElement',] }]
};
if (false) {
    /**
     * Options for NgSelect plugin
     * @type {?}
     * @protected
     */
    NoLiveSearchComponent.prototype._options;
    /**
     * Options for NgSelect plugin
     * @type {?}
     */
    NoLiveSearchComponent.prototype.options;
    /**
     * Current value of live search
     * @type {?}
     */
    NoLiveSearchComponent.prototype.searchValue;
    /**
     * Occurs when current value of live search changes
     * @type {?}
     */
    NoLiveSearchComponent.prototype.searchValueChange;
    /**
     * View child that represents live search element
     * \@internal
     * @type {?}
     */
    NoLiveSearchComponent.prototype.liveSearchElementChild;
    /** @type {?} */
    NoLiveSearchComponent.prototype.ngSelectPlugins;
    /** @type {?} */
    NoLiveSearchComponent.prototype.pluginElement;
    /**
     * @type {?}
     * @private
     */
    NoLiveSearchComponent.prototype._document;
}
