// @ng/sample/dist/es2015/src/components/option/optgroup.component.js
import * as ɵngcc0 from '@angular/core';
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Input, EventEmitter, QueryList, ViewChildren } from "@angular/core";
import { OptionComponent } from "./option.component";
/**
 * Component used for options group in select component
 * @template TValue
 */
export class OptGroupComponent {
    constructor() {
        //######################### private fields #########################
        //######################### private fields #########################
        /**
         * Instance of event emitter for optionsChange
         */
        this._optionsChange = new EventEmitter();
    }
    //######################### public properties #########################
    /**
     * Options assigned to this options group
     * @return {?}
     */
    get options() {
        return this.optionsChildren.toArray();
    }
    /**
     * Occurs when options in this group change
     * @return {?}
     */
    get optionsChange() {
        return this._optionsChange;
    }
    //######################### public methods - implementation of AfterViewInit #########################
    /**
     * Called when view was initialized
     * @return {?}
     */
    ngAfterViewInit() {
        this.optionsChildren.changes.subscribe((/**
         * @return {?}
         */
        () => {
            this._optionsChange.emit();
        }));
    }
}
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(OptGroupComponent, [{
        type: Component,
        args: [{
                selector: 'ng-select>optgroup',
                template: '',
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return []; }, { constructor: [], _optionsChange: [], options: [], optionsChange: [], ngAfterViewInit: [], optionsChildren: [{
            type: ViewChildren,
            args: [OptionComponent]
        }], text: [{
            type: Input
        }] });
OptGroupComponent.ngComponentDef = ɵngcc0.ɵdefineComponent({ type: OptGroupComponent, selectors: [["optgroup"]], factory: function OptGroupComponent_Factory(t) { return new (t || OptGroupComponent)(); }, viewQuery: function OptGroupComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵviewQuery(OptionComponent, true, null);
    } if (rf & 2) {
        var _t;
        (ɵngcc0.ɵqueryRefresh(_t = ɵngcc0.ɵloadViewQuery()) && (ctx.optionsChildren = _t));
    } }, inputs: { text: "text" }, consts: 0, vars: 0, template: function OptGroupComponent_Template(rf, ctx) { }, encapsulation: 2, changeDetection: 0 });
OptGroupComponent.propDecorators = {
    optionsChildren: [{ type: ViewChildren, args: [OptionComponent,] }],
    text: [{ type: Input }]
};
if (false) {
    /**
     * Instance of event emitter for optionsChange
     * @type {?}
     * @private
     */
    OptGroupComponent.prototype._optionsChange;
    /**
     * Children options in this group
     * \@internal
     * @type {?}
     */
    OptGroupComponent.prototype.optionsChildren;
    /**
     * Text that is displayed for this options group
     * @type {?}
     */
    OptGroupComponent.prototype.text;
}
