// @ng/authentication/dist/es2015/src/common/authenticationServiceOptions.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/**
 * Options for authentication service
 * @record
 * @template TUserInfo
 */
export function AuthenticationServiceOptions() { }
if (false) {
    /**
     * Method logs user into system
     * @param {?} accessToken Access token holding authentication information
     * @return {?} Observable
     */
    AuthenticationServiceOptions.prototype.login = function (accessToken) { };
    /**
     * Gets indication whether current state of app is displaying auth page
     * @return {?} boolean
     */
    AuthenticationServiceOptions.prototype.isAuthPage = function () { };
    /**
     * Methods logs out user out of system
     * @return {?} Observable
     */
    AuthenticationServiceOptions.prototype.logout = function () { };
    /**
     * Gets information about user
     * @return {?} Observable
     */
    AuthenticationServiceOptions.prototype.getUserIdentity = function () { };
    /**
     * Redirects current page to authentication page
     * @return {?}
     */
    AuthenticationServiceOptions.prototype.showAuthPage = function () { };
    /**
     * Redirects current page to access denied page
     * @return {?}
     */
    AuthenticationServiceOptions.prototype.showAccessDenied = function () { };
}
/**
 * Authentication options token used for injecting into authentication service
 * @type {?}
 */
export const AUTHENTICATION_SERVICE_OPTIONS = new InjectionToken("AuthenticationServiceOptions");
