// @ng/sample/dist/es2015/src/modules/ngSelect.module.js
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgSelectValuePipe } from '../pipes/ngSelectValue.pipe';
import { OptGroupComponent } from '../components/option/optgroup.component';
import { OptionComponent } from '../components/option/option.component';
import { NgCommonComponent } from '../components/ngCommon/ngCommon.component';
/**
 * Module for select and its options
 */
export class NgSelectModule {
}
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(NgSelectModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    OptionComponent,
                    OptGroupComponent,
                    NgSelectValuePipe,
                    NgCommonComponent
                ],
                exports: [
                    OptionComponent,
                    OptGroupComponent,
                    NgSelectValuePipe,
                    NgCommonComponent
                ]
            }]
    }], null, null);
ɵngcc0.ɵsetComponentScope(OptionComponent, [OptionComponent,
    OptGroupComponent,
    NgCommonComponent, ɵngcc1.NgClass, ɵngcc1.NgComponentOutlet, ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet, ɵngcc1.NgStyle, ɵngcc1.NgSwitch, ɵngcc1.NgSwitchCase, ɵngcc1.NgSwitchDefault, ɵngcc1.NgPlural, ɵngcc1.NgPluralCase], [NgSelectValuePipe, ɵngcc1.AsyncPipe, ɵngcc1.UpperCasePipe, ɵngcc1.LowerCasePipe, ɵngcc1.JsonPipe, ɵngcc1.SlicePipe, ɵngcc1.DecimalPipe, ɵngcc1.PercentPipe, ɵngcc1.TitleCasePipe, ɵngcc1.CurrencyPipe, ɵngcc1.DatePipe, ɵngcc1.I18nPluralPipe, ɵngcc1.I18nSelectPipe, ɵngcc1.KeyValuePipe]);
ɵngcc0.ɵsetComponentScope(NgCommonComponent, [OptionComponent,
    OptGroupComponent,
    NgCommonComponent, ɵngcc1.NgClass, ɵngcc1.NgComponentOutlet, ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet, ɵngcc1.NgStyle, ɵngcc1.NgSwitch, ɵngcc1.NgSwitchCase, ɵngcc1.NgSwitchDefault, ɵngcc1.NgPlural, ɵngcc1.NgPluralCase], [NgSelectValuePipe, ɵngcc1.AsyncPipe, ɵngcc1.UpperCasePipe, ɵngcc1.LowerCasePipe, ɵngcc1.JsonPipe, ɵngcc1.SlicePipe, ɵngcc1.DecimalPipe, ɵngcc1.PercentPipe, ɵngcc1.TitleCasePipe, ɵngcc1.CurrencyPipe, ɵngcc1.DatePipe, ɵngcc1.I18nPluralPipe, ɵngcc1.I18nSelectPipe, ɵngcc1.KeyValuePipe]);
NgSelectModule.ngModuleDef = ɵngcc0.ɵdefineNgModule({ type: NgSelectModule, declarations: function () { return [OptionComponent,
        OptGroupComponent,
        NgSelectValuePipe,
        NgCommonComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [OptionComponent,
        OptGroupComponent,
        NgSelectValuePipe,
        NgCommonComponent]; } });
NgSelectModule.ngInjectorDef = ɵngcc0.defineInjector({ factory: function NgSelectModule_Factory(t) { return new (t || NgSelectModule)(); }, providers: [], imports: [[
            CommonModule
        ],
        [
            OptionComponent,
            OptGroupComponent,
            NgSelectValuePipe,
            NgCommonComponent
        ]] });
