// @ng/sample/dist/es5/src/components/ngCommon/ngCommon.component.js
import * as ɵngcc0 from '@angular/core';
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy } from '@angular/core';
const _c0 = [4, "ngIf"];
const _c1 = [4, "ngFor", "ngForOf"];
const _c2 = [3, "ngClass"];
function NgCommonComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵelementStart(0, "div");
    ɵngcc0.ɵtext(1, "conditional div");
    ɵngcc0.ɵelementEnd();
} }
function NgCommonComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵelementStart(0, "div");
    ɵngcc0.ɵtext(1);
    ɵngcc0.ɵelementEnd();
} if (rf & 2) {
    const itm_r5 = ctx.$implicit;
    ɵngcc0.ɵflushHooksUpTo(1);
    ɵngcc0.ɵtextBinding(1, ɵngcc0.ɵinterpolation1("itm: '", itm_r5, "'"));
} }

/**
 * Component with ngIf ngFor and ngClass
 */
var NgCommonComponent = /** @class */ (function () {
    function NgCommonComponent() {
        this.condition = true;
        this.arrayObj = ['prva', 'druha', 'tretia'];
        this.ngClassObj = { btn: true, "btn-primary": true };
    }
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(NgCommonComponent, [{
        type: Component,
        args: [{
                selector: 'component-sample',
                template: "<div *ngIf=\"condition\">conditional div</div>\r\n<div *ngFor=\"let itm of arrayObj\">itm: '{{itm}}'</div>\r\n<button [ngClass]=\"ngClassObj\"></button>",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return []; }, { condition: [], arrayObj: [], ngClassObj: [] });
NgCommonComponent.ngComponentDef = ɵngcc0.ɵdefineComponent({ type: NgCommonComponent, selectors: [["component-sample"]], factory: function NgCommonComponent_Factory(t) { return new (t || NgCommonComponent)(); }, consts: 3, vars: 3, template: function NgCommonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵtemplate(0, NgCommonComponent_div_0_Template, 2, 0, "div", _c0);
        ɵngcc0.ɵtemplate(1, NgCommonComponent_div_1_Template, 2, 1, "div", _c1);
        ɵngcc0.ɵelement(2, "button", _c2);
    } if (rf & 2) {
        ɵngcc0.ɵelementProperty(0, "ngIf", ɵngcc0.ɵbind(ctx.condition));
        ɵngcc0.ɵflushHooksUpTo(1);
        ɵngcc0.ɵelementProperty(1, "ngForOf", ɵngcc0.ɵbind(ctx.arrayObj));
        ɵngcc0.ɵflushHooksUpTo(2);
        ɵngcc0.ɵelementProperty(2, "ngClass", ɵngcc0.ɵbind(ctx.ngClassObj));
    } }, encapsulation: 2, changeDetection: 0 });
    return NgCommonComponent;
}());
export { NgCommonComponent };
if (false) {
    /** @type {?} */
    NgCommonComponent.prototype.condition;
    /** @type {?} */
    NgCommonComponent.prototype.arrayObj;
    /** @type {?} */
    NgCommonComponent.prototype.ngClassObj;
}
