// @ng/authentication/dist/es2015/src/suppressAuthInterceptor/suppressAuthInterceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { IgnoredInterceptorsService } from '@ng/common';
import { Observable } from 'rxjs';
import { catchError } from 'rxjs/operators';
/**
 * SuppressAuthInterceptor used for intercepting http responses and suppressing 401, 403 statuses
 */
export class SuppressAuthInterceptor {
    //######################### constructors #########################
    /**
     * @param {?} _ignoredInterceptorsService
     */
    constructor(_ignoredInterceptorsService) {
        this._ignoredInterceptorsService = _ignoredInterceptorsService;
    }
    //######################### public methods - implementation of HttpInterceptor #########################
    /**
     * Intercepts http request
     * @param {?} req Request to be intercepted
     * @param {?} next Next middleware that can be called for next processing
     * @return {?}
     */
    intercept(req, next) {
        return next.handle(req).pipe(catchError((/**
         * @param {?} err
         * @return {?}
         */
        (err) => {
            return (/** @type {?} */ (Observable.create((/**
             * @param {?} observer
             * @return {?}
             */
            (observer) => {
                //client error, not response from server, or is ignored
                if (err.error instanceof Error ||
                    (this._ignoredInterceptorsService && this._ignoredInterceptorsService.isIgnored(SuppressAuthInterceptor, req))) {
                    observer.error(err);
                    observer.complete();
                    return;
                }
                //if auth error
                if (err.status == 403 || err.status == 401) {
                    observer.complete();
                    return;
                }
                //other errors
                observer.error(err);
                observer.complete();
            }))));
        })));
    }
}
SuppressAuthInterceptor.decorators = [
    { type: Injectable }
];
/** @nocollapse */
SuppressAuthInterceptor.ctorParameters = () => [
    { type: IgnoredInterceptorsService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    SuppressAuthInterceptor.prototype._ignoredInterceptorsService;
}
/**
 * Provider for proper use of SuppressAuthInterceptor, use this provider to inject this interceptor
 * @type {?}
 */
export const SUPPRESS_AUTH_INTERCEPTOR_PROVIDER = {
    provide: HTTP_INTERCEPTORS,
    multi: true,
    useClass: SuppressAuthInterceptor
};
