// @ng/select/dist/es2015/src/misc/plugin.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * NgSelect plugin interface
 * @record
 */
export function NgSelectPlugin() { }
if (false) {
    /**
     * NgSelect plugin instances available for this plugin
     * @type {?}
     */
    NgSelectPlugin.prototype.ngSelectPlugins;
    /**
     * Element that represents plugin
     * @type {?}
     */
    NgSelectPlugin.prototype.pluginElement;
    /**
     * Options for NgSelect plugin
     * @type {?}
     */
    NgSelectPlugin.prototype.options;
    /**
     * Initialize plugin, to be ready to use, initialize communication with other plugins
     * @return {?}
     */
    NgSelectPlugin.prototype.initialize = function () { };
    /**
     * Initialize plugin options, all operations required to be done with plugin options are handled here
     * @return {?}
     */
    NgSelectPlugin.prototype.initOptions = function () { };
    /**
     * Explicitly runs invalidation of content (change detection)
     * @return {?}
     */
    NgSelectPlugin.prototype.invalidateVisuals = function () { };
}
/**
 * NgSelect plugin generic interface
 * @record
 * @template TOptions
 */
export function NgSelectPluginGeneric() { }
if (false) {
    /**
     * Options for ng select plugin
     * @type {?}
     */
    NgSelectPluginGeneric.prototype.options;
}
/**
 * All available types of plugins for NgSelect
 * @record
 */
export function NgSelectPluginTypes() { }
if (false) {
    /**
     * Handles keyboard events
     * @type {?|undefined}
     */
    NgSelectPluginTypes.prototype.keyboardHandler;
    /**
     * Component used for displaying normal state of select
     * @type {?|undefined}
     */
    NgSelectPluginTypes.prototype.normalState;
    /**
     * Component used for displaying available options for selection
     * @type {?|undefined}
     */
    NgSelectPluginTypes.prototype.popup;
    /**
     * Handles correct position of pop component
     * @type {?|undefined}
     */
    NgSelectPluginTypes.prototype.positioner;
    /**
     * Component used for displaying readonly/disabled state of select, can be null, in that case normal state component is used
     * @type {?|undefined}
     */
    NgSelectPluginTypes.prototype.readonlyState;
    /**
     * Handles obtaining and setting value of component
     * @type {?|undefined}
     */
    NgSelectPluginTypes.prototype.valueHandler;
    /**
     * Contains component that is used for live searching in options
     * @type {?|undefined}
     */
    NgSelectPluginTypes.prototype.liveSearch;
    /**
     * Used for obtaining texts that are displayed in NgSelect
     * @type {?|undefined}
     */
    NgSelectPluginTypes.prototype.textsLocator;
}
/**
 * Base options for every plugin
 * @record
 */
export function PluginOptions() { }
/**
 * Base options for every visual plugin (component)
 * @record
 * @template TCssClasses
 */
export function VisualPluginOptions() { }
if (false) {
    /**
     * Css classes applied to visual plugin (component), possible to override only part of classes
     * @type {?|undefined}
     */
    VisualPluginOptions.prototype.cssClasses;
}
/**
 * Defines interface, that describes minimal set of parameters for specifying plugin for NgSelect
 * @record
 * @template PluginType
 */
export function PluginDescription() { }
if (false) {
    /**
     * Type of plugin that will be dynamically instantiated
     * @type {?|undefined}
     */
    PluginDescription.prototype.type;
    /**
     * Options that will be passed to dynamically instantiated plugin
     * @type {?|undefined}
     */
    PluginDescription.prototype.options;
    /**
     * Optional callback used for obtaining dynamic instance of plugin (allows direct communication with plugin)
     * @type {?|undefined}
     */
    PluginDescription.prototype.instanceCallback;
}
