// @ng/select/dist/es2015/src/plugins/textsLocator/no/noTextsLocator.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, ElementRef, EventEmitter, Inject, Optional } from "@angular/core";
import { NG_SELECT_PLUGIN_INSTANCES } from '../../../components/select/types';
/**
 * Component used for rendering no texts locator
 */
export class NoTextsLocatorComponent {
    //######################### constructor #########################
    /**
     * @param {?} ngSelectPlugins
     * @param {?} pluginElement
     */
    constructor(ngSelectPlugins, pluginElement) {
        this.ngSelectPlugins = ngSelectPlugins;
        this.pluginElement = pluginElement;
        /**
         * Indication that texts should be obtained again, because they have changed
         */
        this.textsChange = new EventEmitter();
    }
    //######################### public methods - implementation of TextsLocator #########################
    /**
     * Gets text for specified key
     * @param {?} key
     * @return {?}
     */
    getText(key) {
        return key;
    }
    /**
     * Initialize plugin, to be ready to use, initialize communication with other plugins
     * @return {?}
     */
    initialize() {
    }
    /**
     * Initialize plugin options, all operations required to be done with plugin options are handled here
     * @return {?}
     */
    initOptions() {
    }
    /**
     * Explicitly runs invalidation of content (change detection)
     * @return {?}
     */
    invalidateVisuals() {
    }
}
NoTextsLocatorComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-no-texts-locator',
                template: '',
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
NoTextsLocatorComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [NG_SELECT_PLUGIN_INSTANCES,] }, { type: Optional }] },
    { type: ElementRef }
];
if (false) {
    /**
     * Options for NgSelect plugin
     * @type {?}
     */
    NoTextsLocatorComponent.prototype.options;
    /**
     * Indication that texts should be obtained again, because they have changed
     * @type {?}
     */
    NoTextsLocatorComponent.prototype.textsChange;
    /** @type {?} */
    NoTextsLocatorComponent.prototype.ngSelectPlugins;
    /** @type {?} */
    NoTextsLocatorComponent.prototype.pluginElement;
}
