// @ng/sample/dist/es5/src/components/option/optgroup.component.js
import * as ɵngcc0 from '@angular/core';
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Input, EventEmitter, QueryList, ViewChildren } from "@angular/core";
import { OptionComponent } from "./option.component";
/**
 * Component used for options group in select component
 * @template TValue
 */
var OptGroupComponent = /** @class */ (function () {
    function OptGroupComponent() {
        //######################### private fields #########################
        //######################### private fields #########################
        /**
         * Instance of event emitter for optionsChange
         */
        this._optionsChange = new EventEmitter();
    }
    Object.defineProperty(OptGroupComponent.prototype, "options", {
        //######################### public properties #########################
        /**
         * Options assigned to this options group
         */
        get: 
        //######################### public properties #########################
        /**
         * Options assigned to this options group
         * @return {?}
         */
        function () {
            return this.optionsChildren.toArray();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OptGroupComponent.prototype, "optionsChange", {
        /**
         * Occurs when options in this group change
         */
        get: /**
         * Occurs when options in this group change
         * @return {?}
         */
        function () {
            return this._optionsChange;
        },
        enumerable: true,
        configurable: true
    });
    //######################### public methods - implementation of AfterViewInit #########################
    /**
     * Called when view was initialized
     */
    //######################### public methods - implementation of AfterViewInit #########################
    /**
     * Called when view was initialized
     * @return {?}
     */
    OptGroupComponent.prototype.ngAfterViewInit = 
    //######################### public methods - implementation of AfterViewInit #########################
    /**
     * Called when view was initialized
     * @return {?}
     */
    function () {
        var _this = this;
        this.optionsChildren.changes.subscribe((/**
         * @return {?}
         */
        function () {
            _this._optionsChange.emit();
        }));
    };
    OptGroupComponent.propDecorators = {
        optionsChildren: [{ type: ViewChildren, args: [OptionComponent,] }],
        text: [{ type: Input }]
    };
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(OptGroupComponent, [{
        type: Component,
        args: [{
                selector: 'ng-select>optgroup',
                template: '',
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return []; }, { _optionsChange: [], options: [], optionsChange: [], ngAfterViewInit: [], optionsChildren: [{
            type: ViewChildren,
            args: [OptionComponent]
        }], text: [{
            type: Input
        }] });
OptGroupComponent.ngComponentDef = ɵngcc0.ɵdefineComponent({ type: OptGroupComponent, selectors: [["optgroup"]], factory: function OptGroupComponent_Factory(t) { return new (t || OptGroupComponent)(); }, viewQuery: function OptGroupComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵviewQuery(OptionComponent, true, null);
    } if (rf & 2) {
        var _t;
        (ɵngcc0.ɵqueryRefresh(_t = ɵngcc0.ɵloadViewQuery()) && (ctx.optionsChildren = _t));
    } }, inputs: { text: "text" }, consts: 0, vars: 0, template: function OptGroupComponent_Template(rf, ctx) { }, encapsulation: 2, changeDetection: 0 });
    return OptGroupComponent;
}());
export { OptGroupComponent };
if (false) {
    /**
     * Instance of event emitter for optionsChange
     * @type {?}
     * @private
     */
    OptGroupComponent.prototype._optionsChange;
    /**
     * Children options in this group
     * \@internal
     * @type {?}
     */
    OptGroupComponent.prototype.optionsChildren;
    /**
     * Text that is displayed for this options group
     * @type {?}
     */
    OptGroupComponent.prototype.text;
}
