// @ng/sample/dist/es5/src/components/option/option.component.js
import * as ɵngcc0 from '@angular/core';
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Input } from "@angular/core";
/**
 * Component used for options in select component
 * @template TValue
 */
var OptionComponent = /** @class */ (function () {
    function OptionComponent() {
        //######################### public properties - template bindings #########################
        //######################### public properties - template bindings #########################
        /**
         * Indication whether is item active
         */
        this.active = false;
        /**
         * Indication whether this option is selected
         */
        this.selected = false;
        //######################### public properties #########################
        /**
         * If specified this option will be displayed in group
         */
        this.group = null;
    }
    OptionComponent.propDecorators = {
        value: [{ type: Input }],
        text: [{ type: Input }]
    };
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(OptionComponent, [{
        type: Component,
        args: [{
                selector: 'ng-select option',
                template: '',
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return []; }, { active: [], selected: [], group: [], value: [{
            type: Input
        }], text: [{
            type: Input
        }] });
OptionComponent.ngComponentDef = ɵngcc0.ɵdefineComponent({ type: OptionComponent, selectors: [["option"]], factory: function OptionComponent_Factory(t) { return new (t || OptionComponent)(); }, inputs: { value: "value", text: "text" }, consts: 0, vars: 0, template: function OptionComponent_Template(rf, ctx) { }, encapsulation: 2, changeDetection: 0 });
    return OptionComponent;
}());
export { OptionComponent };
if (false) {
    /**
     * Indication whether is item active
     * @type {?}
     */
    OptionComponent.prototype.active;
    /**
     * Indication whether this option is selected
     * @type {?}
     */
    OptionComponent.prototype.selected;
    /**
     * Value that will be used if this option will be selected
     * @type {?}
     */
    OptionComponent.prototype.value;
    /**
     * Text that is displayed if this value is selected
     * @type {?}
     */
    OptionComponent.prototype.text;
    /**
     * If specified this option will be displayed in group
     * @type {?}
     */
    OptionComponent.prototype.group;
}
