// @ng/common/dist/es5/src/interceptors/progressInterceptor/progressInterceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { HttpEventType, HTTP_INTERCEPTORS } from '@angular/common/http';
import { tap } from 'rxjs/operators';
import { ProgressIndicatorService } from '../../components/progressIndicator/progressIndicator.service';
import { IgnoredInterceptorsService } from '../../services/ignoredInterceptors/ignoredInterceptors.service';
/**
 * ProgressInterceptor used for intercepting http requests and displaying progress indicatior
 */
var ProgressInterceptor = /** @class */ (function () {
    //######################### constructors #########################
    function ProgressInterceptor(_indicatorSvc, _ignoredInterceptorsService) {
        this._indicatorSvc = _indicatorSvc;
        this._ignoredInterceptorsService = _ignoredInterceptorsService;
    }
    //######################### public methods - implementation of HttpInterceptor #########################
    /**
     * Intercepts http request
     * @param req Request to be intercepted
     * @param next Next middleware that can be called for next processing
     */
    //######################### public methods - implementation of HttpInterceptor #########################
    /**
     * Intercepts http request
     * @param {?} req Request to be intercepted
     * @param {?} next Next middleware that can be called for next processing
     * @return {?}
     */
    ProgressInterceptor.prototype.intercept = 
    //######################### public methods - implementation of HttpInterceptor #########################
    /**
     * Intercepts http request
     * @param {?} req Request to be intercepted
     * @param {?} next Next middleware that can be called for next processing
     * @return {?}
     */
    function (req, next) {
        var _this = this;
        return next.handle(req)
            .pipe(tap((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            if (_this._ignoredInterceptorsService && _this._ignoredInterceptorsService.isIgnored(ProgressInterceptor, req)) {
                return;
            }
            //request started
            if (event.type == HttpEventType.Sent) {
                _this._indicatorSvc.showProgress();
            }
            //response received
            else if (event.type == HttpEventType.Response) {
                _this._indicatorSvc.hideProgress();
            }
        }), (/**
         * @return {?}
         */
        function () { return _this._indicatorSvc.hideProgress(); })));
    };
    ProgressInterceptor.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ProgressInterceptor.ctorParameters = function () { return [
        { type: ProgressIndicatorService },
        { type: IgnoredInterceptorsService, decorators: [{ type: Optional }] }
    ]; };
    return ProgressInterceptor;
}());
export { ProgressInterceptor };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ProgressInterceptor.prototype._indicatorSvc;
    /**
     * @type {?}
     * @private
     */
    ProgressInterceptor.prototype._ignoredInterceptorsService;
}
/**
 * Progress interceptor provider
 * @type {?}
 */
export var PROGRESS_INTERCEPTOR_PROVIDER = {
    provide: HTTP_INTERCEPTORS,
    useClass: ProgressInterceptor,
    multi: true
};
